File: 542-01-matrix/01-matrix.cpp
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int n = mat.size();
        int m = mat[0].size();

        int dr[4] = {0,1,0,-1};
        int dc[4] = {1,0,-1,0};

        vector<vector<int>> vis(n, vector<int>(m, -1));

        queue<vector<int>> q;

        for(int i = 0; i<n; i++){
            for(int j = 0; j<m; j++){
                if(mat[i][j] == 0){
                    vis[i][j] = 0;
                    q.push({i, j, 0});
                }
            }
        }

        while(!q.empty()){
            vector<int> node = q.front();
            q.pop();

            int r = node[0];
            int c = node[1];
            int d = node[2];


            for(int i = 0; i<4; i++){
                int nr = r+dr[i];
                int nc = c+dc[i];

                if(nr>=0 && nr<n && nc>=0 && nc<m && vis[nr][nc]==-1){
                    vis[nr][nc] = d+1;
                    q.push({nr, nc, d+1});
                }
            }
        }

        return vis;
    }
};

File: 542-01-matrix/01-matrix.py
from collections import deque

class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        n = len(mat)
        m = len(mat[0])
        vis = [[-1 for _ in range(m)]for _ in range(n)]
        queue = deque()

        for i in range(n):
            for j in range(m):
                if mat[i][j] == 0:
                    vis[i][j] = 0
                    queue.append([i,j])

        while len(queue):
            r, c = queue.popleft()

            dr = (0,1,0,-1)
            dc = (1,0,-1,0)

            for i in range(4):
                nr = r + dr[i]
                nc = c + dc[i]

                if nr>=0 and nc>=0 and nr<n and nc<m and vis[nr][nc] == -1:
                    vis[nr][nc] = vis[r][c] + 1
                    queue.append([nr,nc])

        return vis



